packages/type-safe-api/docs/developer_guides/type-safe-api/integrations.md

```diff
@@ -47,6 +47,31 @@ When you instantiate your generated `Api` construct, you will need to provide an
 
 For integrating an API operation with a lambda, use `Integrations.lambda(yourLambdaFunction)`.
 
+## S3 integration
+
+For integrating an API operation with an S3 bucket, use `Integrations.s3(s3IntegrationProps)`.
+
+With:
+
+```ts
+s3IntegrationProps = {
+    bucket: yourS3Bucket,
+
+    // Optional props
+
+    // A role to grant access to the S3 bucket to API Gateway, a role with required access is created by default
+    role: yourS3BucketRole
+
+    // The method to use to interact with the S3 bucket, default to integration method
+    method: 'get'
+
+    // The path to use to interact with the S3 bucket, default to integration path
+    path: 'example.json'
+}
+```
```

Comment:
Ooh just a thought for the docs, it'd be great to structure it in 2 sections, a code example in all 3 supported languages for the easy default case, then a code example for all 3 for a more complex custom case.

Here's something you can copy/paste and customise! :) I want to show off this cool feature in the docs haha!

````md
## S3 integration

For integrating an API operation with an S3 bucket, use `Integrations.s3`.

=== "TS"

    ```ts
    Integrations.s3({ bucket: yourS3Bucket });
    ```

=== "Java"

    ```java
    Integrations.s3(S3IntegrationProps.builder()
            .bucket(yourS3Bucket)
            .build());
    ```

=== "Python"

    ```python
    Integrations.s3(bucket=your_s3_bucket)
    ```

!!!note
    By default, this will request an S3 object with the HTTP method and path defined in your API, for example:`GET /pets/{petId}` would perform a `GetObject` for `s3://<yourS3Bucket>/pets/{petId}`, where the path parameter `{petId}` is substituted by the user supplied value.

For more information, for example how the HTTP methods map to S3 operations, refer to the [Tutorial: Create a REST API as an Amazon S3 proxy in API Gateway](https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html).

### Customise S3 Integrations

You can customise the integration to access objects in S3 at a different key to the API operation path. You can also reference path parameters in the overridden path.

For example, you could have an API operation `POST /pets/{petId}/delete`, but map it to an S3 `DeleteObject` operation for a different object key:

=== "TS"

    ```ts
    Integrations.s3({
        bucket: yourS3Bucket,
        method: "delete",
        path: "my-pets/{petId}/details.json"
    });
    ```

=== "Java"

    ```java
    Integrations.s3(S3IntegrationProps.builder()
            .bucket(yourS3Bucket)
            .method("delete")
            .path("my-pets/{petId}/details.json")
            .build());
    ```

=== "Python"

    ```python
    Integrations.s3(
        bucket=your_s3_bucket,
        method="delete",
        path="my-pets/{petId}/details.json"
    )
    ```

````

Reply:
Actually it's also worth noting about customising the error responses too! :)

Reply:
````md
### Error Behaviour

By default, if the S3 operation throws an error (for example an object at the requested path does not exist), these will be mapped to an empty `500` error response.

You can customise the error behaviour by passing an `errorIntegrationResponse`.

* `ErrorIntegrationResponses.catchAll()` - default behaviour, returns a `500` response for any S3 error
* `ErrorIntegrationResponses.none()` - TODO @valebedu what happens when S3 errors with this? haha
* `ErrorIntegrationResponses.custom(...)` - custom error responses, mapping an S3 HTTP status regex to the corresponding response API gateway should return
````